src/ecs/rendering/systems.rs

```diff
@@ -1,81 +1,59 @@
 //! Rendering system.
 
-use renderer::Config as DisplayConfig;
-use renderer::Rgba;
-use renderer::prelude::*;
-use winit::EventsLoop;
-
 use assets::BoxedErr;
-use ecs::{Fetch, Join, ReadStorage, System, World};
-use ecs::SystemExt;
-use ecs::rendering::components::*;
+use ecs::{Fetch, System, SystemExt, World};
 use ecs::rendering::resources::{Factory, AmbientColor};
 use ecs::transform::components::*;
 use error::{Error, Result};
+//use renderer::prelude::*;
+use renderer::{Config as DisplayConfig, Renderer, Rgba, Light, Mesh, Material};
+use renderer::pipe::{Pipeline, PipelineData, PipelineBuild};
+
+use winit::EventsLoop;
 
 /// Rendering system.
 #[derive(Derivative)]
 #[derivative(Debug)]
-pub struct RenderSystem {
-    pipe: Pipeline,
+pub struct RenderSystem<P> {
+    pipe: P,
     #[derivative(Debug = "ignore")]
     renderer: Renderer,
-    scene: Scene,
 }
 
-impl<'a> System<'a> for RenderSystem {
-    type SystemData = (Fetch<'a, Camera>,
-     Fetch<'a, Factory>,
-     Fetch<'a, AmbientColor>,
-     ReadStorage<'a, Transform>,
-     ReadStorage<'a, LightComponent>,
-     ReadStorage<'a, MaterialComponent>,
-     ReadStorage<'a, MeshComponent>);
+impl<'a, P> System<'a> for RenderSystem<P>
+    where P: Pipeline,
+{
+    type SystemData = (
+        Fetch<'a, Factory>,
+        <P as PipelineData<'a>>::Data,
+    );
 
     fn run(
         &mut self,
-        (camera, factory, ambient_color, globals, lights, materials, meshes): Self::SystemData,
+        (factory, data): Self::SystemData,
     ) {
         use std::time::Duration;
 
         while let Some(job) = factory.jobs.try_pop() {
             job.exec(&mut self.renderer.factory);
         }
 
-        self.scene.clear();
-
-        for (mesh, material, global) in (&meshes, &materials, &globals).join() {
-            self.scene.add_model(Model {
-                material: material.0.clone(),
-                mesh: mesh.as_ref().clone(),
-                pos: global.0.into(),
-            });
-        }
-
-        self.scene.set_ambient_color(ambient_color.0.clone());
-
-        for light in lights.join() {
-            self.scene.add_light(light.0.clone());
-        }
-
-        self.scene.add_camera(camera.clone());
-
-        self.renderer.draw(
-            &self.scene,
-            &self.pipe,
-            Duration::from_secs(0),
-        );
+        self.renderer.draw(&mut self.pipe, data, Duration::from_secs(0));
     }
 }
 
-impl<'a, 'b> SystemExt<'a, (&'b EventsLoop, PipelineBuilder, Option<DisplayConfig>)>
-    for RenderSystem {
+impl<'a, B, P> SystemExt<'a, (&'a EventsLoop, B, Option<DisplayConfig>)>
+    for RenderSystem<P>
+    where B: PipelineBuild<Pipeline=P>,
+          P: Pipeline,
+{
     /// Create new `RenderSystem`
```

Comment:
Can't you remove the P bound here, and just do `for RenderSystem<B::Pipeline> where B: PipelineBuild`?